NotificationHubs/src/com/windowsazure/messaging/RetryUtil.java

```diff
@@ -72,18 +72,46 @@ public static RetryPolicy getRetryPolicy(RetryOptions options) {
                 }
 
                 if (error instanceof TimeoutException) {
-                	System.out.println("TimeoutException error occurred. Retrying operation. Attempt: " + attempt);
-                	
+                	System.out.println("TimeoutException error occurred. Retrying operation. Attempt: " + attempt); 
                     return retryPolicy.calculateRetryDelay(error, attempt);
-                } else if (error instanceof QuotaExceededException && (((QuotaExceededException) error).getIsTransient())) {
+                } else if (error instanceof QuotaExceededException) {
                 	System.out.println("Retryable error occurred. Retrying operation. Attempt: " + attempt + ". Error: " + error);
-                	
                     return retryPolicy.calculateRetryDelay(error, attempt);
+                } else if (error instanceof NotificationHubsException) {                    	
+                    	int statusCode = ((NotificationHubsException) error).getHttpStatusCode();
+                    	if (statusCode == 408 || statusCode == 500 || statusCode == 503 || statusCode == 504) {
+                    		System.out.println("Retryable error occurred. Retrying operation. Attempt: " + attempt + ". Error: " + error);
+                    		return retryPolicy.calculateRetryDelay(error, attempt);
+                    	}
+                    	
+                    	System.out.println("Error is not a TimeoutException nor is it a retryable exception." + error);
+                        throw Exceptions.propagate(error);
                 } else {
                 	System.out.println("Error is not a TimeoutException nor is it a retryable exception." + error);
                     throw Exceptions.propagate(error);
                 }
             })
             .flatMap(Mono::delay);
     }
+
+    public static Optional<Integer> parseRetryAfter(HttpResponse response)
```

Comment:
IMHO, this method is essentially adding its own non-idiomatic Java error handling. I'd much prefer that the contract here is that this method return a number of seconds (or milliseconds) to delay, and if there's failure to just throw the exception.

Reply:
Looks like this is Reactor way to retry something. We need to throw some exception and retry some method if the exception was thrown.  Should we change it?

Reply:
Maybe a good compromise is to make it package-protected instead of public. I don't want to expose this type in our public surface area if it comes from an external dependency, but I recognize it may be the right way to do this for Reactor.